### PR TITLE
implement submenu "settings" to the XCSoar-based menu

### PR DESCRIPTION
**Pull request by Blaubart**
I have added to the XCSoar-based menu the submenu for settings. Now the rotation and the brightness of the display, the language, the timeout for the autostart, as well as the activation or deactivation of SSH, variod and sensord can be made.

### CODE REVIEW

File: src/OV/OpenVarioMenu.cpp [hunk 1]
```diff
@@ -119,15 +119,87 @@ FileMenuWidget::Prepare([[maybe_unused]] ContainerWindow &parent,
...
+};
+
+void
+ScreenRotationWidget::Prepare([[maybe_unused]] ContainerWindow &parent,
+                              [[maybe_unused]] const PixelRect &rc) noexcept
+{
+  AddButton("Lanscape", [](){
+    static constexpr const char *argv[] = {
+      "/bin/sh", "-c", 
+      "sed -ir 's/^[#]*\s*rotation=.*/rotation=0/' /boot/config.uEnv && echo The screen was rotated to landscape",
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp:147]: What's the point of all those `echo` calls?
If you want to display a message to the user, you should use C++ code, not shell code.

File: src/OV/OpenVarioMenu.cpp [hunk 2]
```diff
@@ -227,16 +934,16 @@ SystemMenuWidget::Prepare([[maybe_unused]] ContainerWindow &parent,
...
 
-    RunProcessDialog(UIGlobals::GetMainWindow(),
-                     UIGlobals::GetDialogLook(),
-                     "Update Maps", argv);
+     RunProcessDialog(UIGlobals::GetMainWindow(),
+                      UIGlobals::GetDialogLook(),
+                      "Update Maps", argv);
   });
-
+
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp:946]: Lots and lots of whitespace changes

File: src/OV/OpenVarioMenu.cpp [hunk 3]
```diff
@@ -353,7 +1064,7 @@ MainMenuWidget::Prepare([[maybe_unused]] ContainerWindow &parent,
 
     TWidgetDialog<FileMenuWidget>
       sub_dialog(WidgetDialog::Full{}, dialog.GetMainWindow(),
-                 GetLook(), "OpenVario File");
+                 GetLook(), "OpenVario Files");
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp:1067]: Don't hide unrelated changes in this commit - if you wish to change labels, that should be a new commit

File: src/OV/OpenVarioMenu.cpp [hunk 4]
```diff
@@ -107,7 +107,7 @@ FileMenuWidget::Prepare([[maybe_unused]] ContainerWindow &parent,
                      UIGlobals::GetDialogLook(),
                      "Downloading files", argv);
   });
-
+
```
[the remark this answers is not in the document]
Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp:110]: Sorry, is changed

File: src/OV/OpenVarioMenu.cpp (hunk 2 above)
Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp:946]: Unnecessary spaces as in line 946 have been eliminated

File: src/OV/OpenVarioMenu.cpp (hunk 3 above)
Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp]: OK

File: src/OV/OpenVarioMenu.cpp (hunk 1 above)
Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp:147]: OK, I will change it soon

Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp:147]: Can you give me a hint, witch function I have to use?
I think I have to use the function like:
PrettyFunctionNameHere(&dialog, "Here is your text");

**Comment by lordfolken** [comment]
Hey @Blaubart !
3 things:
 * commit messages structure: 
    component: short commit message 80chars
     Longer explanation is necessary

We have over 20 years of commits. They are an awesome resource why things have been designed and why decisions have been taken. Its a message in a bottle to your future self or others that may follow. So please treat your work with a succinct  description. :-)
      
* commit history:
     please rebase your code before pushing and avoid merges..
* Whitespace unnecessary changes, commit hiding etc:
   This makes things very hard to read and understand. 
    There is an awesome tool called precommit: https://pre-commit.com/ that checks your commit for blank lines, retabbing code formatting etc before you push.

Thanks!

**Comment by MaxKellermann** [comment]
PR contains merge commit and CI test fails.

File: src/OV/OpenVarioMenu.cpp (hunk 1 above)
Comment (Scumi) [reply src/OV/OpenVarioMenu.cpp:147]: I suggest to add `echo "0" >/sys/class/graphics/fbcon/rotate_all` instead of the textual feedback. This rotates the screen immediately without the need of a reboot. And if the user sees the new orientation directly, you don't need to bother him with additional feedback (especially pop-up dialogues which are often rather annoying).

File: src/OV/OpenVarioMenu.cpp (hunk 2 above)
Comment (lordfolken) [reply src/OV/OpenVarioMenu.cpp]: The other lines as well. You can use https://pre-commit.com/ as a githook. Here is an example file: https://github.com/XCSoar/mapgen/blob/master/.pre-commit-config.yaml
It will check your commits and eliminate white space etc. before committing.

File: src/OV/OpenVarioMenu.cpp (hunk 1 above)
Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp:147]: Thanks a lot!!

Comment (MaxKellermann) [reply src/OV/OpenVarioMenu.cpp:147]: Don't write files by spawning the shell.

>  But for other echo messages I need to know which function is responsible for display messages with c++ code

The message "The screen was rotated to landscape" seems rather pointless, doesn't it? Just don't display it. The user can see the effect without you telling him.

Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp:147]: > The message "The screen was rotated to landscape" seems rather pointless, doesn't it? Just don't display it. The 
> user can see the effect without you telling him.

At the moment, rotating the screen only becomes active after a reboot, because only then is the config loaded. With Scumi's suggestion, the screen would be rotated directly. I'll test this with echo first, and then figure out how to do it without a shell

Comment (MaxKellermann) [reply src/OV/OpenVarioMenu.cpp:147]: > figure out how to do it without a shell

https://github.com/XCSoar/XCSoar/blob/cdd8108cd60aa1fff2f6830002fd27cc320ff0ff/src/Kobo/System.cpp#L412

Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp:147]: Thanks!!!

Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp:147]: It works! I can write strings to files. But I have one problem.

The string is always inserted in the first line. In the "/boot/config.uEnv" file, for example, there is more information, such as: 
rotation=0
LANG=de_DE.UTF-8
timeout=10

I still have to figure out how to find the right line and completely replace the old string with the new one. Can the File::WriteExisting function do this?

**Comment by MaxKellermann** [comment]
Bad commit messages. If I see "Update <filename>", I won't even start reviewing it because I know the diff is going to be bad.

**Comment by Blaubart** [comment]
I wil start a new PR witch has a better structure!!

**Comment by MaxKellermann** [comment]
No, please don't. There is never a reason to do that, it will only harm the review workflow.

File: build/ov.mk [hunk 5]
```diff
@@ -6,6 +6,7 @@ OV_MENU_SOURCES = \
 	$(SRC)/Hardware/CPU.cpp \
 	$(SRC)/Hardware/DisplayDPI.cpp \
 	$(SRC)/Hardware/RotateDisplay.cpp \
+	$(SRC)/Hardware/DisplayGlue.cpp \
```
Comment (MaxKellermann) [inline build/ov.mk:9]: Why is this "necessary"?

File: src/OV/OpenVarioMenu.cpp [hunk 6]
```diff
@@ -414,9 +414,7 @@ int main()
 
   switch (action) {
   case LAUNCH_SHELL:
-    execl("/bin/bash", "bash", "--login", nullptr);
-    execl("/bin/ash", "-ash", nullptr);
-    execl("/bin/ash", "-sh", nullptr);
+    execl("/bin/sh", "sh", "-c", "clear; bash -il", nullptr);
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp:1194]: .... why?? Your commit message does not explain anything.

File: src/OV/OpenVarioMenu.cpp [hunk 7]
```diff
@@ -410,6 +410,14 @@ Main()
 
 int main()
 {
+  /*the x-menu is waiting a second to solve timing problem with display rotation */
+  std::chrono::high_resolution_clock hrc;
+  auto start = hrc.now();
+  while(std::chrono::duration_cast<std::chrono::milliseconds>(hrc.now() - start).count() < 1000)
+  {
+    //I'm just waiting ;-)
+  }
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp:1188]: Oh nooooooooo not a busy loop! It's not the 80ies, these days, you must never busy-loop.

File: src/OV/OpenVarioMenu.cpp [hunk 8]
```diff
@@ -391,6 +391,7 @@ Main()
 
   UI::TopWindowStyle main_style;
   main_style.Resizable();
+  main_style.InitialOrientation(Display::DetectInitialOrientation());
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp:1162]: I wish I could cherry-pick this commit, but it fails to compile. Sigh.

File: src/OV/OpenVarioMenu.cpp [hunk 9]
```diff
@@ -328,6 +328,17 @@ MainMenuWidget::Prepare([[maybe_unused]] ContainerWindow &parent,
...
+  AddButton("Logbook", [this](){
+    CancelTimer();
+    static constexpr const char *argv[] = {
+      "/usr/bin/logbook.sh", nullptr
+    };
+
+    RunProcessDialog(UIGlobals::GetMainWindow(),
+                     UIGlobals::GetDialogLook(),
+                     "Logbook", argv);
+  });
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp:1097]: This is not acceptable and will never get merged, as I told you already several months ago.

File: src/OV/OpenVarioMenu.cpp [hunk 10]
```diff
@@ -315,7 +315,14 @@ class MainMenuWidget final
...
-    return RowFormWidget::KeyPress(key_code);
+
+  /* ignore escape key at first menu page */
+    if (key_code != KEY_ESCAPE) {
+	    return RowFormWidget::KeyPress(key_code);
+	}
+	else {
+        return true;
+	}
   }
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp:1076]: Indentation is completely broken.

File: src/OV/OpenVarioMenu.cpp [hunk 11]
```diff
@@ -304,8 +304,13 @@ class MainMenuWidget final
...
     RowFormWidget::Show(rc);
 
-    if (remaining_seconds > 0)
+    if (remaining_seconds > 0) {
       ScheduleTimer();
+    }
+    else {
+	  HideRow(Controls::TIMER);
+      StartXCSoar();
+	}
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp:1058]: Indentation broken, and I don't understand the commit message.

File: src/OV/OpenVarioMenu.cpp [hunk 12]
```diff
@@ -264,14 +266,15 @@ class MainMenuWidget final
...
-  unsigned remaining_seconds = 3;
-
 public:
   MainMenuWidget(UI::Display &_display, UI::EventQueue &_event_queue,
                  WndForm &_dialog) noexcept
     :RowFormWidget(_dialog.GetLook()),
      display(_display), event_queue(_event_queue),
-     dialog(_dialog) {}
+     dialog(_dialog)
+     {
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp:1017]: Broken indentation

File: src/OV/OpenVarioMenu.cpp [hunk 13]
```diff
@@ -240,6 +240,8 @@ SystemMenuWidget::Prepare([[maybe_unused]] ContainerWindow &parent,
 class MainMenuWidget final
   : public RowFormWidget
 {
+  int remaining_seconds = 3;
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp:984]: Bad type

File: src/OV/OpenVarioMenu.cpp [hunk 14]
```diff
@@ -207,16 +207,6 @@ SystemMenuWidget::Prepare([[maybe_unused]] ContainerWindow &parent,
...
 
-  AddButton("Update Maps", [](){
-    static constexpr const char *argv[] = {
-      "/usr/bin/update-maps.sh", nullptr
-    };
-
-    RunProcessDialog(UIGlobals::GetMainWindow(),
-                     UIGlobals::GetDialogLook(),
-                     "Update Maps", argv);
-  });
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp]: Wait, this removes a feature. This is confusing. If it belongs somewhere else, then this commit should move it, and not delete it.

File: src/OV/OpenVarioMenu.cpp [hunk 15]
```diff
@@ -197,6 +197,18 @@ void
 SystemMenuWidget::Prepare([[maybe_unused]] ContainerWindow &parent,
                           [[maybe_unused]] const PixelRect &rc) noexcept
 {
+  AddButton("WiFi Settings", [](){
+    static constexpr const char *argv[] = {
+      "/bin/sh", "-c", 
+      "printf '\nWiFi-Settings are not implemented, yet!! \n\nIf you are interessted to help with this, write me an email: dirk@freevario.de'", 
+      nullptr
+    };
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp]: shell script not acceptable. Never will be. Will never be merged.

File: src/OV/OpenVarioMenu.cpp [hunk 16]
```diff
@@ -105,11 +105,14 @@ class SystemMenuWidget final
...
+
 public:
   SystemMenuWidget(UI::Display &_display, UI::EventQueue &_event_queue,
-                   const DialogLook &look) noexcept
-    :RowFormWidget(look),
-     display(_display), event_queue(_event_queue) {}
+          WndForm &_dialog) noexcept
+    :RowFormWidget(_dialog.GetLook()),
+     display(_display), event_queue(_event_queue),
+     dialog(_dialog) {}
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp]: I don't understand this commit, but I know for sure that indentation is broken, again.

File: src/OV/OpenVarioMenu.cpp [hunk 17]
```diff
@@ -99,6 +99,101 @@ FileMenuWidget::Prepare([[maybe_unused]] ContainerWindow &parent,
...
+};
+
+void
+ScreenSSHWidget::Prepare([[maybe_unused]] ContainerWindow &parent,
+                         [[maybe_unused]] const PixelRect &rc) noexcept
+{
+  AddButton("Enable", [](){
+    static constexpr const char *argv[] = {
+      "/bin/sh", "-c", 
+      "systemctl enable dropbear.socket && printf '\nSSH has been enabled'",
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp:624]: No shell scripts acceptable. Will never be merged.

File: src/OV/OpenVarioMenu.cpp [hunk 18]
```diff
@@ -99,6 +99,115 @@ FileMenuWidget::Prepare([[maybe_unused]] ContainerWindow &parent,
...
+                             [[maybe_unused]] const PixelRect &rc) noexcept
+
+{
+  AddButton("immediately", [this](){
+    SaveTimeout(0);
+        static constexpr const char *argv[] = {
+      "/bin/sh", "-c", 
+      "echo Automatic timeout was set to 0s", 
+      nullptr
+    };
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp:526]: WTF, you wrote a shell script and all it does is display a message to be displayed by the menu???

File: src/OV/OpenVarioMenu.cpp [hunk 19]
```diff
@@ -99,6 +99,73 @@ FileMenuWidget::Prepare([[maybe_unused]] ContainerWindow &parent,
...
+  void Prepare(ContainerWindow &parent,
+               const PixelRect &rc) noexcept override;
+
+void SaveBrightness(const string &brightness);
+};
+
+void
+ScreenBrightnessWidget::SaveBrightness(const string &brightness)
+{
+    File::WriteExisting(Path("/sys/class/backlight/lcd/brightness"), (brightness).c_str());
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp:266]: Bad indent

File: src/OV/OpenVarioMenu.cpp [hunk 20]
```diff
@@ -99,6 +99,73 @@ FileMenuWidget::Prepare([[maybe_unused]] ContainerWindow &parent,
   });
 }
 
+class ScreenBrightnessWidget final
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp:243]: Unused class. Commits which add an unused class will not be merged.

File: src/OV/OpenVarioMenu.cpp [hunk 21]
```diff
@@ -99,6 +99,60 @@ FileMenuWidget::Prepare([[maybe_unused]] ContainerWindow &parent,
   });
 }
 
+class ScreenRotationWidget final
```
Comment (MaxKellermann) [inline src/OV/OpenVarioMenu.cpp:189]: Unused class. Will not be merged.

**Comment by MaxKellermann** [comment]
I stopped reading (and commenting) commits because I had enough; I'll come back here only after you fix all of my complaints.

File: build/ov.mk (hunk 5 above)
Comment (Blaubart) [reply build/ov.mk:9]: otherwise I get an error during compiling XCSoar:
|   LINK    output/UNIX/bin/OpenVarioMenu
| src/OV/OpenVarioMenu.cpp:175: error: undefined reference to 'Display::DetectInitialOrientation()'
| src/OV/OpenVarioMenu.cpp:1162: error: undefined reference to 'Display::DetectInitialOrientation()'
| collect2: error: ld returned 1 exit status
| make: *** [build/ov.mk:38: output/UNIX/bin/OpenVarioMenu] Error 1
| ERROR: oe_runmake failed

File: src/OV/OpenVarioMenu.cpp (hunk 6 above)
Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp:1194]: one of the lines is enough to call the shell

File: src/OV/OpenVarioMenu.cpp (hunk 9 above)
Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp:1097]: I removed again

File: src/OV/OpenVarioMenu.cpp (hunk 10 above)
Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp:1076]: fixed

File: src/OV/OpenVarioMenu.cpp (hunk 11 above)
Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp:1058]: fixed

File: src/OV/OpenVarioMenu.cpp (hunk 12 above)
Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp:1017]: fixed

File: src/OV/OpenVarioMenu.cpp (hunk 19 above)
Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp:266]: hope it is better now

File: src/OV/OpenVarioMenu.cpp (hunk 17 above)
Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp:624]: how can I enable or disable a daemon like SSH or variod without systemctl using C++ code? systemctl is a shell command

File: src/OV/OpenVarioMenu.cpp (hunk 8 above)
Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp:1149]: It fails because ov.mk needs:
$(SRC)/Hardware/DisplayGlue.cpp \

File: src/OV/OpenVarioMenu.cpp (hunk 13 above)
Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp]: changed it from int to unsigned

File: src/OV/OpenVarioMenu.cpp (hunk 14 above)
Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp]: I reverted two commits and make a new one. Now submenu "Update Maps" is moved to the submenu "Files"

File: src/OV/OpenVarioMenu.cpp (hunk 8 above)
Comment (Scumi) [reply src/OV/OpenVarioMenu.cpp:1149]: But that is not enough. I cherry-picked that commit and can't get it to compile either, even after adding almost the whole content of `main.mk`. The dependencies of `DisplayGlue.cpp` blow up the makefile and I wonder if there is a "cleaner" way of doing this.

Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp:1149]: Hmmm, I can compile the Image for the OpenVario without any problems.

Comment (Scumi) [reply src/OV/OpenVarioMenu.cpp:1149]: But you can't compile the binary for the OpenVarioMenu standalone (try `make output/UNIX/bin/OpenVarioMenu`) and that is what's required. The dependencies are probably taken from the preceding XCSoar build.

Comment (Blaubart) [reply src/OV/OpenVarioMenu.cpp:1149]: thank you for your explanation!!

**Comment by MaxKellermann** [comment]
This PR is getting even more messed up now.

**Comment by MaxKellermann** [comment]
Since it looks like you are not anymore interested in working on getting this PR merged into XCSoar, I'm closing this PR now.

**Comment by MaxKellermann** [comment]
btw. @Blaubart, your repository contains commits that you claim authorship for, but the code was really written by me. For example: https://github.com/Blaubart/XCSoar/commit/0ee4809868af01892095c980e070155562ad8b4f
Don't do that!

**Comment by Blaubart** [comment]
I claim no authorship for the code. I added lines to your code to make the menu usable for me. And I uploaded these changes. I don't know where I claim something in any way.

**Comment by MaxKellermann** [comment]
The "Author" field of that commit adding code I wrote is your name.

**Comment by Blaubart** [comment]
can't find any author field

**Comment by MaxKellermann** [comment]
```
commit 0ee4809868af01892095c980e070155562ad8b4f
Author: Blaubart <Blaubart@gmx.de>
Date:   Wed Dec 20 22:40:55 2023 +0100

    including x-menu
```

**Comment by Blaubart** [comment]
Yes, I am the author of the commit, but not necessarily the code. That's a subtle difference. But please don't waste our time. Be helpful and say specifically how I want to change the owner of the commit to you, and of course I will do it!

**Comment by MaxKellermann** [comment]
> Yes, I am the author of the commit, but not necessarily the code. That's a subtle difference.

You are wrong here, probably because you have no idea what you are talking about. The (not-so) "subtle" difference is between "committer" and "author".

The "committer" is the "author of the commit", in your words.
But the "author" is the author of the code. There's no other valid interpretation.

You claim to be both. I asked you not to claim authorship of code I wrote, so....

Claiming authorship of somebody else's work is copyright violation, by the way. Be careful with that.

> But please don't waste our time. Be helpful and say specifically how I want to change the owner of the commit to you, and of course I will do it!

`git commit --amend --author=.....` but I guess you can read manpages, too..?
